File: app/modules/ytdlp/android/src/test/java/expo/modules/ytdlp/platforms/InstagramTest.kt
package expo.modules.ytdlp.platforms

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/** Mirrors server/src/platforms/Instagram.test.ts. */
class InstagramTest {
    private fun instagram(url: String = "https://www.instagram.com/p/abc123/") = Instagram(url)

    // --- isRetryableError ---

    @Test
    fun `isRetryableError is false for a slideshow-carousel post with no video`() {
        assertFalse(instagram().isRetryableError(RuntimeException("ERROR: There is no video in this post")))
    }

    @Test
    fun `isRetryableError is true for any other error`() {
        assertTrue(instagram().isRetryableError(RuntimeException("HTTP Error 403: Forbidden")))
    }

    // --- describeError ---

    @Test
    fun `describeError rewrites a no-video error to the slideshow-not-supported key`() {
        val error = RuntimeException("There is no video in this post")
        assertEquals("errors.instagramSlideshowNotSupported" to null, instagram().describeError(error))
    }

    @Test
    fun `describeError matches the no-video pattern case-insensitively`() {
        val error = RuntimeException("NO VIDEO FORMATS FOUND")
        assertEquals("errors.instagramSlideshowNotSupported" to null, instagram().describeError(error))
    }

    @Test
    fun `describeError wraps an unrelated message as errors raw with the raw text as a param`() {
        assertEquals("errors.raw" to mapOf("raw" to "boom"), instagram().describeError(RuntimeException("boom")))
    }
}

File: app/__tests__/server/platforms/Instagram.test.ts
import { Instagram } from "../../../server/platforms/Instagram.js";

function instagram(url = "https://www.instagram.com/p/abc123/") {
  return new Instagram(url);
}

describe("isRetryableError", () => {
  it("is false for a slideshow/carousel post with no video", () => {
    expect(instagram().isRetryableError(new Error("ERROR: There is no video in this post"))).toBe(false);
  });

  it("is true for any other error", () => {
    expect(instagram().isRetryableError(new Error("HTTP Error 403: Forbidden"))).toBe(true);
  });
});

describe("describeError", () => {
  it("rewrites a no-video error to the slideshow-not-supported key", () => {
    expect(instagram().describeError(new Error("There is no video in this post"))).toEqual({
      key: "errors.instagramSlideshowNotSupported",
    });
  });

  it("matches the no-video pattern case-insensitively", () => {
    expect(instagram().describeError(new Error("NO VIDEO FORMATS FOUND"))).toEqual({
      key: "errors.instagramSlideshowNotSupported",
    });
  });

  it("passes through a non-matching message as the raw key's param", () => {
    expect(instagram().describeError(new Error("network timeout"))).toEqual({
      key: "errors.raw",
      params: { raw: "network timeout" },
    });
  });

  it("uses the fallback as the raw param when err isn't an Error instance", () => {
    expect(instagram().describeError("boom", "fallback")).toEqual({
      key: "errors.raw",
      params: { raw: "fallback" },
    });
  });
});
